### ISSUE TITLE
Loosing data during forecast process

### ISSUE THREAD

**Issue by lg1000** [business-science/modeltime#133]
Hi,

was just trying to do some simple forecasting with modeltime packages.  After I pass the data through modeltime_forecast(), there are only 48 of 240 original variables left in the resultset. I posted the question with examples on stackoverflow. Please let me know, if I am doing something wrong with your package syntax.

https://stackoverflow.com/questions/69193491/how-to-handle-forecast-data-melt-and-unmelt-generated-by-modeltime-predictio

Best Regards 
Leo

**Comment by mdancho84** [comment]
I think it's a matter of understanding how best to forecast with modeltime. 

First, you are using `rsample::initial_time_split()`, which doesn't respect groups to my knowledge. I can tell this from your train/test split cv plan. 

``` r
split_obj <- initial_time_split(data, prop = 0.8)  

split_obj %>%
    tk_time_series_cv_plan() %>%
    plot_time_series_cv_plan(dates, value)
```

![image](https://user-images.githubusercontent.com/13734662/133459601-ee53c96c-d422-40ea-a828-5a798d40abba.png)


So I'd recommend switching over to `timetk::time_series_split()`. It looks better if I do this. 

``` r 
split_obj_2 <- time_series_split(data, assess = "1 year", cumulative = TRUE)

split_obj_2 %>%
    tk_time_series_cv_plan() %>%
    plot_time_series_cv_plan(dates, value)
``` 

Next, if I run calibration and then visualize, I can see what the forecasts look like on our test set for a sample of 9 of the series. This tells me that I'm making progress. Personally, I'd try a different model like xgboost at this point because I can see Prophet isn't the greatest at global modeling. But, if you are happy, then continue. 

``` r
## calibration
calibration_tbl_prophet <- models_tbl_prophet %>%
    modeltime_calibrate(new_data = testing(split_obj_2))

## forecast
fc_prophet <- calibration_tbl_prophet %>%
    modeltime_forecast(
        new_data = testing(split_obj_2),
        actual_data = data,
        keep_data = TRUE
    ) 

fc_prophet %>%
    filter(as.numeric(variable) %in% 1:9) %>%
    group_by(variable) %>%
    plot_modeltime_forecast(.facet_ncol = 3)

```

![image](https://user-images.githubusercontent.com/13734662/133460534-495e31aa-f698-4856-9c3a-0ac61d682d7b.png)


Then, we can make a future forecast say "1 year" into the future. Again, we should check our forecast results by plotting a sample of 9 time series or so. 

``` r
# Make a future forecast

refit_tbl_prophet <- calibration_tbl_prophet %>%
    modeltime_refit(data = data)

future_fc_prophet <- refit_tbl_prophet %>%
    modeltime_forecast(
        new_data = data %>% group_by(variable) %>% future_frame(.length_out = "1 year"),
        actual_data = data,
        keep_data = TRUE
    )

future_fc_prophet %>%
    filter(as.numeric(variable) %in% 1:9) %>%
    group_by(variable) %>%
    plot_modeltime_forecast(.facet_ncol = 3)
```

![image](https://user-images.githubusercontent.com/13734662/133461361-aeec4006-5019-4192-99de-6104b6823e4a.png)


Finally, reformat the data into the structure that you need. 

``` r
future_wide_tbl <- future_fc_prophet %>%
    filter(.key == "prediction") %>%
    select(.model_id, .model_desc, dates, variable, .value) %>%
    pivot_wider(
        id_cols     = c(.model_id, .model_desc, dates),
        names_from  = variable, 
        values_from = .value
    )

future_wide_tbl[names(df)]
```

![image](https://user-images.githubusercontent.com/13734662/133462131-cc600f60-c7ae-4a07-8f68-158f0f085eee.png)


This produces the next 1 year or data, which looks like 241 observations.

**Comment by mdancho84** [comment]
For further information on how to forecast with `modeltime`, I highly recommend my [High Performance Time Series Forecasting Course](https://university.business-science.io/p/ds4b-203-r-high-performance-time-series-forecasting/). This will save you countless hours of debugging issues like these. Plus I have a private slack channel that is more appropriate for questions like these.

**Comment by lg1000** [comment]
Thanks a lot for the detailed response! I will close the issue, as the lost data problem is solved

**Comment by mdancho84** [comment]
Great. Please mark the stack overflow response as the solution too so others can easily know that it was resolved.

**Comment by lg1000** [comment]
Just marked your stackoverflow response. I left a comment concerning the forecast output. Here it appears to be always the same output, also when I change the dataframe to the actual data of my costumer or random numbers. I did another approach, where I was pulling the data and the tidymodels through a loop, where everything works fine, but that solution is too slow to go live

**Comment by mdancho84** [comment]
Yes, I recommend that you use a better algorithm- try xgboost with the time series signature features. This is one of the algorithms I teach in my course for global modeling. It's lighting fast and can be quite accurate for 1000s of time series. Much better than prophet for this task.

**Comment by lg1000** [comment]
ok, this looks pretty bad for prophet, while on my real data it wasn`t performing too bad at all. In the end I want to take a mixture of models and evaluate the forecasts using Theils U2 and MSE decomposition, benchmarked against some costumers hand made forecast. Even without any tuning, the tidymodels are performing three times as well. I am just testing the arima with boost. It is awfully slow compared to prophet so far, but I guess I just need to figure out the right learning rate

**Comment by lg1000** [comment]
In the end, I still cannot figure out, why the forecast results of different variables are exactly alike. To illustrate I will append a code with random numbers. As you can see in the result "future_wide_tbl", there are a few varying forecasts in the beginning, but then it produces the same forecast again and again: 
```

`suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(lubridate))
suppressPackageStartupMessages(library(tidymodels))
suppressPackageStartupMessages(library(modeltime))
suppressPackageStartupMessages(library(timetk))

# create mostly random data for illustration
date <- ymd("2016-01-01")+ months(0:59)
randVec <- function(ii){
	r1 <- c(sample(0:4, 4, replace = TRUE) , 3,
			sample(1:9, 0, replace = TRUE), 
			sample(1:9, 0, replace = TRUE), 
			sample(2:9, 0, replace = TRUE), 4,
			sample(1:6, 1, replace = TRUE), 3,
			sample(2:7, 0, replace = TRUE), 
			sample(3:6, 0, replace = TRUE),
			sample(2:5, 1, replace = TRUE),
			sample(3:6, 1, replace = TRUE),
			sample(3:5, 1, replace = TRUE), 
			sample(1:3, 0, replace = TRUE), 3,
			sample(0:4, 4, replace = TRUE) , 3,
			sample(1:9, 0, replace = TRUE), 
			sample(1:9, 0, replace = TRUE), 
			sample(2:9, 0, replace = TRUE), 4,
			sample(1:6, 1, replace = TRUE), 3,
			sample(2:11, 0, replace = TRUE), 
			sample(3:11, 0, replace = TRUE),
			sample(5:13, 1, replace = TRUE),
			sample(3:14, 1, replace = TRUE),
			sample(3:16, 1, replace = TRUE), 
			sample(7:17, 0, replace = TRUE), 3,
			sample(0:4, 4, replace = TRUE) , 3,
			sample(1:9, 0, replace = TRUE), 
			sample(1:9, 0, replace = TRUE), 
			sample(2:9, 0, replace = TRUE), 4,
			sample(1:6, 1, replace = TRUE), 3,
			sample(2:11, 0, replace = TRUE), 
			sample(3:17, 0, replace = TRUE),
			sample(5:25, 1, replace = TRUE),
			sample(3:47, 1, replace = TRUE),
			sample(3:44, 1, replace = TRUE), 
			sample(7:33, 0, replace = TRUE), 3,
			sample(0:4, 4, replace = TRUE) , 3,
			sample(1:9, 0, replace = TRUE), 
			sample(1:9, 0, replace = TRUE), 
			sample(2:9, 0, replace = TRUE), 4,
			sample(1:6, 1, replace = TRUE), 3,
			sample(2:11, 0, replace = TRUE), 
			sample(3:17, 0, replace = TRUE),
			sample(5:25, 1, replace = TRUE),
			sample(3:47, 1, replace = TRUE),
			sample(3:44, 1, replace = TRUE), 
			sample(7:33, 0, replace = TRUE), 3,  
			sample(1:9, 0, replace = TRUE), 
			sample(2:9, 0, replace = TRUE), 4,
			sample(1:6, 1, replace = TRUE), 3,
			sample(2:11, 0, replace = TRUE), 
			sample(3:17, 0, replace = TRUE), 
			sample(5:25, 1, replace = TRUE),
			sample(3:47, 1, replace = TRUE),
			sample(3:44, 1, replace = TRUE), 
			sample(7:33, 0, replace = TRUE), 3,
			sample(1:6, 1, replace = TRUE), 3,
			sample(2:11, 0, replace = TRUE), 
			sample(3:17, 0, replace = TRUE),
			sample(5:25, 1, replace = TRUE),
			sample(3:47, 1, replace = TRUE),
			sample(3:44, 1, replace = TRUE), 
			sample(7:33, 0, replace = TRUE)
	)   
}
set.seed(1234)
df <- as.data.frame(replicate(240, randVec()
)
)
df <- bind_cols(df, date) %>%
	rename(c(date = ...241)
	)
df2 <- head(df,n=48)

# melt it down
data <- reshape2::melt(df2, id.var='date')
data %>% as_tibble() -> data
data %>%
	filter(as.numeric(variable) %in% 1:9) %>%
	group_by(variable) %>%
	plot_time_series(date, value, .facet_ncol = 3, .smooth = F)

#split
split_obj_3 <- time_series_split(data, assess = "1 year", cumulative = TRUE)
split_obj_3 %>%
	tk_time_series_cv_plan() %>%
	plot_time_series_cv_plan(date, value)


################################################################## -> fake in sample

# fit model
model_fit_arima_boosted <- arima_boost(
	min_n = 2,
	learn_rate = 0.015
) %>%
	set_engine(engine = "auto_arima_xgboost") %>%
	fit(value ~ date + as.numeric(date) + factor(month(date, label = TRUE), ordered = F),
		data = training(split_obj_3))

# model table
models_tbl_arima_boosted <- modeltime_table(model_fit_arima_boosted)

# calibration
calibration_tbl_arima_boosted <- models_tbl_arima_boosted %>%
	modeltime_calibrate(new_data = testing(split_obj_3)
	)

# fake a future forecast
refit_tbl_arima_boosted <- calibration_tbl_arima_boosted %>%
	modeltime_refit(data = data)

future_fc_arima_boosted <- refit_tbl_arima_boosted %>%
	modeltime_forecast(
		new_data = data %>% group_by(variable) %>% future_frame(.length_out = "1 year"),
		actual_data = data,
		keep_data = TRUE
	)

future_fc_arima_boosted %>%
	filter(as.numeric(variable) %in% 1:9) %>%
	group_by(variable) %>%
	plot_modeltime_forecast(.facet_ncol = 3)

# Reformat as wide
future_wide_tbl <- future_fc_arima_boosted %>%
	filter(.key == "prediction") %>%
	select(.model_id, .model_desc, date, variable, .value) %>%
	pivot_wider(
		id_cols     = c(.model_id, .model_desc, date),
		names_from  = variable, 
		values_from = .value
	)

future_wide_tbl[names(df)]
```
